### ISSUE TITLE
0.8.2: focus_follows_mouse has become unreliable

### ISSUE THREAD

**Issue by Nairou** [herbstluftwm/herbstluftwm#908]
In 0.8.0 this worked perfectly, window focus would follow the mouse instantly. In the 2 days since I updated to 0.8.2, I've had numerous cases where moving the mouse over a window didn't move focus, and I had to move the mouse over another window and back again to trigger the focus change.

There was even one case where mouse focus failed entirely, and I had to actually click on windows to move focus until I restarted herbstluftwm.

Moving focus is normally something I do without thinking, so I tend to notice focus not moving when I try to type in the new window and nothing happens.

**Comment by t-wissmann** [comment]
I have similar issues, but so far I failed reproducing it (It only happens from time to time here). Do you know a way of reproducing it?

I'm sure that this issue was introduced by one of the following pull requests: https://github.com/herbstluftwm/herbstluftwm/pull/868 or https://github.com/herbstluftwm/herbstluftwm/pull/864

**Comment by t-wissmann** [comment]
I think I can reproduce it with two frames and with
```
hc set frame_border_width 20
```
so my first guess is that https://github.com/herbstluftwm/herbstluftwm/pull/864 is the bad guy.

**Comment by Nairou** [comment]
Upping my frame_border_width from 3 to 20 didn't appear to make a difference for me.

It's still intermittent, though it seems more likely if a particular window has had focus for a while. When I try to test it, and move focus away from the browser and back again, I'll see it skip once on the way back (not giving focus back to the browser), but as I continue to move the mouse back and forth between windows quickly it doesn't skip any more.

**Comment by Nairou** [comment]
I think I found a reliable way to reproduce it. It seems to happen if the mouse moves fast enough that it doesn't touch any part of the window frame or border.

If you flick the mouse over fast enough that you can see the trail of cursor movement where it goes directly from the contents of one window to the contents of the next window, without touching anything in between, then focus doesn't follow.

**Comment by t-wissmann** [comment]
Does it make a difference whether the two windows are in the same frame or in different frames?

**Comment by Nairou** [comment]
Just tested with two windows in the same frame, and it does the same thing there as well. Focus only switches if the border gets touched. All other tests were against windows in separate frames.

**Comment by t-wissmann** [comment]
And can you reproduce it if you switch the tag to floating mode?

**Comment by t-wissmann** [comment]
However, it happens much rarer in floating mode (and it also happens in 0.8.0). So I think the issue is a race condition that always existed in hlwm and is now much more likely with the changes in the above mentioned PRs.

What happens internally is, that after windows are tiled on a monitor, all `EnterNotify` events are dropped. The `EnterNotify` event is the event reporting that the mouse cursor entered a window, and this is used in `focus_follows_mouse`. We drop these events after each tiling step because we don't want to focus a window if its moved under the cusor. The race condition is the following:

- the mouse cursor is moved quickly, producing a enter notify for window 1
- while the screen is being tiled, the mouse cursor is moved, producing an enter notify for window 2
- when the tiling is finished, all enter notify events in the event queue are dropped, dropping the enter notify for window 2 (i.e. window 1 stays focused though the cursor is within window 2)

Maybe we can fix this fundamentally by simply not tiling the screen when only the focus changes.

**Comment by Nairou** [comment]
Is it possible to drop all events in the queue matching the window that triggered the tiling, rather than dropping them all?

**Comment by t-wissmann** [comment]
For me, the issue is fixed by the PR #909. Can you give it a try? To test it, you can either replace the herbstluftwm-executable in your system by the one compiled from git, or you can also do the following in your git directory:
```
git checkout racy-enternotify-drop
cd build && cmake ..
make
../valgrind-xephyr.sh
```
This will open a nested session in a xephyr window with the default configuration (the default modifier is <kbd>Alt</kbd>, not <kbd>Mod4</kbd>)

**Comment by Nairou** [comment]
I replaced my running copy, and so far it looks fixed for me as well. No matter how fast I move the mouse, focus moves with it. :+1:

**Comment by t-wissmann** [comment]
Cool, thanks for testing and for your help debugging this issue!

**Comment by blarz** [comment]
This can be closed right? 0.8.3 fixes this

**Comment by t-wissmann** [comment]
Yes, thanks! Fixed by #909 and in v0.8.3
